src/AutoRest.CSharp.V3/Output/Models/Types/ObjectType.cs

```diff
@@ -365,19 +378,53 @@ private IEnumerable<ObjectTypeProperty> BuildProperties()
             }
         }
 
+        // Enumerates all schemas that were merged into this one, excludes the inherited schema
+        private IEnumerable<ObjectSchema> GetCombinedSchemas()
+        {
+            var inherited = EnumerateHierarchy().Select(type => type.Schema).ToHashSet();
+
+            yield return _objectSchema;
+
+            foreach (var parent in _objectSchema.Parents!.All)
+            {
+                if (parent is ObjectSchema objectParent && !inherited.Contains(objectParent))
+                {
+                    yield return objectParent;
+                }
+            }
+        }
 
         private CSharpType? CreateInheritedType()
         {
-            foreach (ComplexSchema complexSchema in _objectSchema.Parents!.Immediate)
+            var sourceBaseType = _sourceTypeMapping?.ExistingType?.BaseType;
+            if (sourceBaseType != null &&
+                sourceBaseType.SpecialType != SpecialType.System_ValueType &&
+                sourceBaseType.SpecialType != SpecialType.System_Object &&
+                _typeFactory.TryCreateType(sourceBaseType, out CSharpType? baseType))
             {
-                if (complexSchema is ObjectSchema parentObjectSchema)
+                return baseType;
+            }
+
+            var objectSchemas = _objectSchema.Parents!.Immediate.OfType<ObjectSchema>().ToArray();
+
+            ObjectSchema? selectedSchema = null;
+
+            foreach (var objectSchema in objectSchemas)
+            {
+                // Take first schema or the one with discriminator
+                if (selectedSchema == null || objectSchema.Discriminator != null)
                 {
-                    CSharpType type = _typeFactory.CreateType(parentObjectSchema, false);
-                    Debug.Assert(!type.IsFrameworkType);
-                    return type;
+                    selectedSchema = objectSchema;
```

Comment:
Really interesting way to write this. However, I don't think it works as it would always select the last `objectSchema` with a discriminator. You'd likely want the first one.

Reply:
You are right. Also this loop is kinda strange

Reply:
Much better. 👍